components/director/internal/domain/formation/service.go

```diff
@@ -1161,3 +1232,72 @@ func setToSlice(set map[string]bool) []string {
+			return errors.Wrapf(updateError, "while updating error state: %s", errors.Wrapf(err, "while sending notification for formation with ID: %q", formation.ID).Error())
+		}
+		notificationErr := errors.Wrapf(err, "while sending notification for formation with ID: %q and name: %q", formation.ID, formation.Name)
+		log.C(ctx).Error(notificationErr)
+		return notificationErr
+	}
+
+	if response.Error != nil && *response.Error != "" {
+		err = s.setFormationToErrorState(ctx, formation, *response.Error, formationassignment.ClientError, errorState)
+		if err != nil {
+			return errors.Wrapf(err, "while updating error state for formation with ID: %q and name: %q", formation.ID, formation.Name)
+		}
+
+		err = errors.Errorf("Received error from formation webhook response: %v", *response.Error)
+		log.C(ctx).Error(err)
+		return err
+	}
+
+	if *response.ActualStatusCode == *response.SuccessStatusCode {
+		formation.State = model.ReadyFormationState
+		log.C(ctx).Infof("Updating formation with ID: %q and name: %q to state: %s", formation.ID, formation.Name, model.ReadyFormationState)
+		if err := s.formationRepository.Update(ctx, formation); err != nil {
+			return errors.Wrapf(err, "while updating formation with ID: %q and name: %q to state: %s", formation.ID, formation.Name, model.ReadyFormationState)
+		}
+	}
+
+	return nil
+}
+
+func (s *service) setFormationToErrorState(ctx context.Context, formation *model.Formation, errorMessage string, errorCode formationassignment.AssignmentErrorCode, state model.FormationState) error {
+	formation.State = state
+
+	formationError := formationassignment.AssignmentErrorWrapper{Error: formationassignment.AssignmentError{
+		Message:   errorMessage,
+		ErrorCode: errorCode,
+	}}
+
+	marshaledErr, err := json.Marshal(formationError)
+	if err != nil {
+		return errors.Wrapf(err, "While preparing error message for formation with ID: %q", formation.ID)
+	}
+	formation.Error = marshaledErr
+
+	if err := s.formationRepository.Update(ctx, formation); err != nil {
+		return err
+	}
+	log.C(ctx).Infof("Formation with ID: %s set to state: %s", formation.ID, formation.State)
+	return nil
+}
+
+func determineFormationState(ctx context.Context, formationTemplateID, formationTemplateName string, formationTemplateWebhooks []*model.Webhook) model.FormationState {
+	if len(formationTemplateWebhooks) == 0 {
+		log.C(ctx).Infof("Formation template with ID: %q and name: %q does not have any webhooks. The formation will be created with %s state", formationTemplateID, formationTemplateName, model.ReadyFormationState)
+		return model.ReadyFormationState
+	}
+
+	if len(formationTemplateWebhooks) == 1 && *formationTemplateWebhooks[0].Mode == model.WebhookModeSync {
+		log.C(ctx).Infof("Formation template with ID: %q and name: %q have one webhook with %s mode. The formation will be created with %s state", formationTemplateID, formationTemplateName, model.WebhookModeSync, model.ReadyFormationState)
+		return model.ReadyFormationState
```

Comment:
We need this because the `determineFormationState` is invoked before the formation creation and we need to know with what state to create it. After formation creation, we generate notifications and the line you mentioned is part of the `processFormationNotifications` which is responsible for setting the correct state based on the executed notification request.
Currently, both places are ready state because we're in the synchronous case. But when we introduce async logic most probably the "first linked state" will change to `INITIAL`. I left it for the async logic because we probably need to take into account other places/flows as well.